### DISCUSSION TITLE
ICC profiles

### DISCUSSION THREAD

**Discussion by telenkor** [jim-easterbrook/python-exiv2#14] [Q&A] [answered]
Please tell me, сan we work with ICC color profiles? I'm interested in adding a profile to image. If this is not possible, please help me understand the situation:
There is a raw picture, let's call it: «_source_raw_».
There is a RGB picture, let's call it: «_rgb_target_».
«_rgb_target_» currently has the ICC profile, but does not have other metadata.
```python
        image = exiv2.ImageFactory.open(rgb_target)
        image.setExifData(exif_from_source_raw)
        image.writeMetadata()
```
This does only one good thing - it adds metadata from _source_raw_ to _rgb_target_, but at the same time it destroys the existing ICC-profile. I understand that it replaces the old with the new. Is it possible to somehow combine existing ICC data and new other data? (e.g. image.setExifData(exif_from_source_raw + exif_from_rgb_target))

**Comment by jim-easterbrook** [comment]
The `Exiv2::Image` class has `iccProfile()` and `setIccProfile()` methods which should be accessible from Python.
https://exiv2.org/doc/classExiv2_1_1Image.html#a7f0c33da949b99bcf0b6c63d47ca6f33

**Reply by telenkor** [reply]
To be honest, I still have difficulty understanding how to translate Exiv2 methods into Python. What happens is mainly thanks to the examples from the description of your project. How I imagine it:
```python
image = exiv2.ImageFactory.open(‘image.jpg’)
www = exiv2.DataBuf() #created an object, but how to place the profile into…
image.setIccProfile(www)
image.writeMetadata()
```
How, for example, would you put the ICCprofile data into a DataBuf and assign it to an image?
P.S. You will forgive me that I may be distracting you from more important activities, but I really still have a hard time understanding this.

**Reply by jim-easterbrook** [reply]
The exiv2 library can't modify the contents of a Python object. It's a C++ library that knows nothing about Python. The python-exiv2 interface "extracts" data from a Python object to pass it to libexiv2, and converts libexiv2 outputs to Python objects.

To get the ICC profile from an image and copy it to another image is easy:
```
icc = image_1.iccProfile()
image_2.setIccProfile(icc)
```

The `Exiv2::DataBuf` type has a `data()` method to get its contents as a Python buffer. You can create a `Exiv2::DataBuf` of the size you want by passing an integer to its 2nd constructor, but you can't then modify its contents. You should be able to pass some data to the 3rd constructor, but this is not currently callable from Python. I'll fix that in the next release of python-exiv2.
https://exiv2.org/doc/structExiv2_1_1DataBuf.html

**Reply by telenkor** [reply]
Unfortunately, it didn't work. Tried many images. With sRGB and others.
Maybe the problem is that the file is searched in the image and not embedded directly? So, by the way, it would be much more convenient.
```python
image = exiv2.ImageFactory.open("0068-crop-1200.tif")
icc = image.iccProfile()
image_2 = exiv2.ImageFactory.open("0670_1.jpg")
image_2.setIccProfile(icc)
image_2.writeMetadata()
```

```
Traceback (most recent call last):
  File "/Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pydev/pydevd.py", line 1500, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Applications/PyCharm CE.app/Contents/plugins/python-ce/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/Volumes/WorkRAID/python/pc_projects/test_3.11/_tmp0002.py", line 43, in <module>
    image_2.setIccProfile(icc)
exiv2.Exiv2Error: Not a valid ICC Profile
```

**Reply by jim-easterbrook** [reply]
Would the ICC profile of a TIF image be valid for JPG? (I don't know anything about ICC profiles, but I have copied them successfully.)

Do you need to call `image.readMetadata()` before getting its ICC profile?

**Reply by telenkor** [reply]
Icc/icm profiles are immutable, cross-platform and therefore identical for all files where they are embedded.
Mmm, I call `image.readMetadata()` before getting its ICC profile. And it had an effect. I was able to successfully take a profile from TIFF and JPEG.
How can I still get a DataBuf object with a profile from its path (e.g. "/Vol…….tr/srgb.icc"), so as not to take it from the pic?
Look, I managed to initiate the FileIO object:
```
sss = exiv2.FileIo("/Volu…/srgb.icc")

<Swig Object of type 'Exiv2::FileIo *' at 0x10c61eb30>
```

But we need to get an object of this type:
```
icc = image.iccProfile()

<Swig Object of type 'Exiv2::DataBuf *' at 0x10b204ff0>
```
However, actions similar to Exiv2::FileIo are not successful:
`sss = exiv2.DataBuf("/Volu…/srgb.icc")` Is wrong and not working.

**Reply by jim-easterbrook** [reply]
As I said earlier:
> You should be able to pass some data to the 3rd constructor, but this is not currently callable from Python. I'll fix that in the next release of python-exiv2.

You can't currently copy data into an `Exiv2::DataBuf`. Please wait a few days for me to do a new release with a fix. I've assigned it to bug #15.

**Reply by jim-easterbrook** [reply]
I've made some more changes to the Python interface to `Exiv2::DataBuf` this morning. In the next release the `data()` method returns a Python `memoryview` of the data which is writeable. This means you will be able to modify ICC profiles in place without needing to create new `exiv2.DataBuf` objects.

**Reply by telenkor** [reply]
Thank you very much! I'll be looking forward to the release of the build on [pypi](https://pypi.org).

**Reply by jim-easterbrook** [reply]
I'm uploading v0.15.0 to PyPI now. I think it has everything you need to work with ICC profiles.

Thank you for reporting the bugs you've found and starting some discussions. They've made me look much closer at how some things work, which has led to some worthwhile improvements in python-exiv2.

**Reply by telenkor** [reply]
Great news at the end of the week! I'll be exploring the new version this weekend. I am glad that I helped make my tiny contribution to the development of the project.

**Comment by telenkor** [comment]
Everything works fine: modification metadata, embedding icc (for Jpeg and Tiff). But the WebP format came as a surprise. Metadata is good. But when embedding an ICC profile, the file becomes unreadable. A check in the HEX editor showed that the EXIV2 does not put down the necessary marker "ICCP" [WebP Container Specification](https://dev.exiv2.org/projects/exiv2/wiki/The_Metadata_in_WEBP_files). Everything is OK on the CLI version 0.27.7 of Windows :confused:

![wrong](https://github.com/jim-easterbrook/python-exiv2/assets/31967374/9a763812-0f6f-4aba-a4dc-689b4ed8ae04)
![good](https://github.com/jim-easterbrook/python-exiv2/assets/31967374/d000afd2-836f-4cd1-a0d3-b1b906a54eeb)

I manually place the desired marker and the image comes to life. Of course, as long as the data is in RAM, I can patch the file. However, this behavior in relation to this format seems strange. Maybe there was an error somewhere or some factor was missed?

Here is a slightly simplified code from my project regarding working with EXIV2. If you comment out the `# Embedd ICC profile` line, everything is fine with the file.

```python
from io import BytesIO

from PIL import Image
import exiv2  # v0.15

color_space = None
# Initialize common ICC profiles (sRGB IEC61966-2.1 and Adobe RGB 1998) in buffer
with open("/tools/srgb.icc", "rb") as file_icc:
    icc_data_srgb = file_icc.read()
with open("/tools/adobe_rgb_1998.icc", "rb") as file_icc:
    icc_data_argb98 = file_icc.read()

input_file = "/Users/test/Desktop/0050-master-flat.tif"
img = Image.open(input_file)
img.thumbnail((900, 900), Image.Resampling.BICUBIC)

buffer_ = BytesIO()
# img.save(buffer_, format="JPEG", subsampling=1, progressive=1, optimize=1, quality=75)
img.save(buffer_, format="webp", method=6, quality=75)

image_ref_file = exiv2.ImageFactory.open(input_file)
image_ref_file.readMetadata()
exif_ref = image_ref_file.exifData()

i = exif_ref.begin()
end = exif_ref.end()
exif_custom = exiv2.ExifData()
while i != end:
    if "ColorSpace" in i.key():
        # 1 - sRGB, 2 - Adobe RGB 1998
        if i.toString() == "1":
            color_space = 1
        elif i.toString() == "2":
            color_space = 2
    if i.key() in (
            'Exif.Photo.ExposureTime',
            'Exif.Photo.ISOSpeedRatings',
            'Exif.Image.Make'
    ):
        exif_custom[i.key()] = i.value()
    next(i)

image = exiv2.ImageFactory.open(buffer_.getvalue())
image.setExifData(exif_custom)

# Embedd ICC profile
image.setIccProfile(exiv2.DataBuf(icc_data_srgb if color_space == 1 else icc_data_argb98))

image.writeMetadata()

io = image.io()
# with open("/Volumes/Disk/out/0050-master-flat.jpg", 'wb') as file:
#     file.write(io.mmap())
with open("/Volumes/Disk/out/0050-master-flat.webp", 'wb') as file:
    file.write(io.mmap())
io.munmap()
io.close()
```
[Tiff and profiles](https://drive.google.com/file/d/1kjzRi47FgSmnzmoZUwrfJlVd-Ug1M3D6/view?usp=sharing)
P.S. If a **_valid_** WebP does not open anywhere, then open it in a browser (Chrome, Mozilla and others).

**Reply by jim-easterbrook** [reply, marked as answer]
That's a bug in the libexiv2 webp implementation: https://github.com/Exiv2/exiv2/issues/2734

When I get libexiv2 0.28.1 (or later) properly working with python-exiv2 then it should be fixed.

**Reply by telenkor** [reply]
This is not a critical issue. I made a patch to fix it. Hopefully the new library design will be usable with python-exiv2 soon.
```python
        # Patch ICCP chunk header https://github.com/Exiv2/exiv2/issues/2734
        image_content = bytearray(io.mmap())
        if self.preview_compression_format == "webp":
            wrong_header = image_content.index(b"\x00\x56\x50\x38\x58")
            right_header = b"\x00\x49\x43\x43\x50"
            image_content[wrong_header:wrong_header + 5] = right_header
```

**Reply by jim-easterbrook** [reply]
Your snippet shows an `mmap()` without a corresponding `munmap()`. It's easier to use the `io` object's buffer interface to initialise the `bytearray`: `image_content = bytearray(io)`.

You might be able to do an in-place modification by using a `memoryview` of the `io` object.
```
        # Patch ICCP chunk header https://github.com/Exiv2/exiv2/issues/2734
        if self.preview_compression_format == "webp":
            with memoryview('io') as ptr:
                wrong_header = bytes(ptr).index(b"\x00\x56\x50\x38\x58")
                right_header = b"\x00\x49\x43\x43\x50"
                ptr[wrong_header:wrong_header + 5] = right_header
        image_content = bytearray(io)
```

**Reply by telenkor** [reply]
Thanks, let's try this.
